aiohttp/web_urldispatcher.py

```diff
@@ -639,59 +639,67 @@ def __iter__(self) -> Iterator[AbstractRoute]:
 
     async def _handle(self, request: Request) -> StreamResponse:
         rel_url = request.match_info["filename"]
+        filename = Path(rel_url)
+        if filename.anchor:
+            # rel_url is an absolute name like
+            # /static/\\machine_name\c$ or /static/D:\path
+            # where the static dir is totally different
+            raise HTTPForbidden()
+
+        unresolved_path = self._directory.joinpath(filename)
+        loop = asyncio.get_event_loop()
+        return await loop.run_in_executor(
+            None, self._resolve_path_to_response, unresolved_path
+        )
+
+    def _resolve_path_to_response(self, unresolved_path: Path) -> StreamResponse:
+        """Takes the unresolved path and queries the file system to form a response."""
+        # Check for access outside the root directory. For follow symlinks, URI
+        # cannot traverse out, but symlinks can. Otherwise, no access outside
+        # root is permitted.
         try:
-            filename = Path(rel_url)
-            if filename.anchor:
-                # rel_url is an absolute name like
-                # /static/\\machine_name\c$ or /static/D:\path
-                # where the static dir is totally different
-                raise HTTPForbidden()
-            unresolved_path = self._directory.joinpath(filename)
             if self._follow_symlinks:
                 normalized_path = Path(os.path.normpath(unresolved_path))
                 normalized_path.relative_to(self._directory)
-                filepath = normalized_path.resolve()
+                file_path = normalized_path.resolve()
             else:
-                filepath = unresolved_path.resolve()
-                filepath.relative_to(self._directory)
-        except (ValueError, FileNotFoundError) as error:
-            # relatively safe
-            raise HTTPNotFound() from error
-        except HTTPForbidden:
-            raise
-        except Exception as error:
-            # perm error or other kind!
-            request.app.logger.exception(error)
+                file_path = unresolved_path.resolve()
+                file_path.relative_to(self._directory)
+        # TODO(PY39): Remove OSError to appease Windows
+        except (ValueError, RuntimeError, OSError) as error:
+            # ValueError for relative check; RuntimeError for circular symlink.
             raise HTTPNotFound() from error
 
-        # on opening a dir, load its contents if allowed
-        if filepath.is_dir():
-            if self._show_index:
-                try:
+        # if path is a directory, return the contents if permitted. Note the
+        # directory check will raise if a segment is not readable.
+        try:
+            if file_path.is_dir():
```

Comment:
Its unfortuate that this will stat the file again since `is_dir` and `is_file` are implemented like `return S_ISREG(self.stat().st_mode)`

It should be cached but still not free.

Reply:
I guess we could reverse this and do the `is_file` check first as it would be the common case and you only end up with an extra `stat` on the dir or not a normal file case.

Reply:
I agree perf could be improved here, but TBH I'd rather keep status quo for this PR.  

I'd rather think about this in a follow up to address [#8182](https://github.com/aio-libs/aiohttp/issues/8182).  Anything I change here would have to be rethought out, where there are weird cases to think about like someone has a directory `foo/` next to a `foo.gz`.

Reply:
Anyone that cares about perf should not be using this in the first place. nginx or similar solution for serving static resources is virtually guaranteed to be the better choice.

Reply:
When we were fixing GHSA-5h86-8mv2-jq9f we found static usage much more common than expected. While more ideal, `nginix` isn't always something that can be deployed because of where the python application sits in the stack.  Millions+ of installs use the static handler every day because of this and it still makes sense to get what we can out of it.

Reply:
That's a fair point @bdraco I would argue that vast majority of those installs do have the option of a better solution, but don't know or care to implement it. I vaguely remember documentation steering people towards it.